vk/response.go

```diff
@@ -1,77 +1,54 @@
 package vk
 
 import (
+	"context"
 	"encoding/json"
 	"net/http"
-
-	"github.com/pkg/errors"
-	"github.com/suborbital/vektor/vlog"
 )
 
-// Response represents a non-error HTTP response
-type Response struct {
-	status int
-	body   interface{}
-}
+type RawString string
+type RawBytes []byte
 
-// Respond returns a filled-in response
-func Respond(status int, body interface{}) Response {
-	r := Response{
-		status: status,
-		body:   body,
+// RespondWeb converts a value to either raw string or json, and sends it to the client. Ctx is a placeholder here, it
+// is currently unused, but will be used for tracing / logging help purposes later.
+func RespondWeb(_ context.Context, w http.ResponseWriter, data any, statusCode int) error {
+	// If there is nothing to marshal then set status code and return.
+	if statusCode == http.StatusNoContent {
+		w.WriteHeader(statusCode)
+		return nil
 	}
 
-	return r
-}
-
-// R is `Respond` for those who prefer terse code
-func R(status int, body interface{}) Response {
-	return Respond(status, body)
-}
-
-// TODO: add convenience helpers for status codes
-
-const (
-	contentTypeJSON        contentType = "application/json"
-	contentTypeTextPlain   contentType = "text/plain"
-	contentTypeOctetStream contentType = "application/octet-stream"
-)
-
-// converts _something_ into bytes, best it can:
-// if data is Response type, returns (status, body processed as below)
-// if bytes, return (200, bytes)
-// if string, return (200, []byte(string))
-// if struct, return (200, json(struct))
-// otherwise, return (500, nil)
-func responseOrOtherToBytes(l *vlog.Logger, data interface{}) (int, []byte, contentType) {
-	if data == nil {
-		return http.StatusNoContent, []byte{}, contentTypeTextPlain
+	// If the data is deliberately set as raw bytes or raw strings, return them as is. If we don't have this, they will
+	// be marshalled into json, which will wrap a naked string into double quotes.
+	switch i := data.(type) {
+	case RawBytes:
+		w.Header().Set("Content-Type", "text/plain")
+		w.WriteHeader(statusCode)
+		_, _ = w.Write(i)
+		return nil
+	case RawString:
```

Comment:
If it were up to me, I would say that every response is a json response, and adjust the test cases / expectations where the expected response is a naked string. That way this isn't even needed.

If we do want to keep naked strings though, I think this is the way to go.

My reasoning is that if we do RespondX for all different formats, we're going to duplicate a lot of code. This way we're still communicating that these two specifically need to be left alone and naked because of their types, but everything else will default to json.

What would a separate RespondX get us that would improve this or make the intent clearer?

Reply:
The trouble with forcing everything to be JSON is that the places where Vektor is used can very likely return non-JSONy things. I'm thinking binary files like images or even weird text formats like XML. At the end of the day of course any user of Vektor could just write to the `rw` themselves and do what they need to do.

Reply:
> The trouble with forcing everything to be JSON is that the places where Vektor is used can very likely return non-JSONy things.

Normally I'd agree, but because calling `vk.RespondWeb` or whatever that returns json is entirely optional from the handler, developers can choose to call their own response function, or even write to the responsewriter directly as needed. No one is forced to use our RespondWeb func because it's been decoupled from the chain.

Reply:
I'm going to break out a RespondJSON, and a RespondRaw that takes a string / byteslice / runeslice.

Reply:
@jagger27 https://github.com/suborbital/vektor/pull/101/commits/0eed113bef2ca9b9ae1437e8f9cb4ec81326956f